File: ingestion-service/src/main/scala/datainsider/ingestion/domain/column/NestedColumn.scala
package datainsider.ingestion.domain

import datainsider.ingestion.domain.column.DefaultExpression

@SerialVersionUID(20200715L)
case class NestedColumn(
    name: String,
    displayName: String,
    description: Option[String] = None,
    nestedColumns: Seq[Column],
    isEncrypted: Boolean = false,
    isNullable: Boolean = false
) extends Column {
  override val defaultExpr: Option[String] = None
  override val defaultExpression: Option[DefaultExpression] = None
  override def copyTo(name: String, displayName: String): Column = this.copy(name = name, displayName = displayName)

  override def copyTo(
      name: String,
      displayName: String,
      description: Option[String],
      isNullable: Boolean,
      defaultExpression: Option[DefaultExpression],
      isEncrypted: Boolean
  ): Column = {
    this.copy(
      name = name,
      displayName = displayName,
      description = description,
      isNullable = isNullable,
      isEncrypted = isEncrypted
    )
  }
}

File: rocket-bi-web/di_core/domain/Model/Column/Implement/NestedColumn.ts
import { Column } from '@core/domain/Model/Column/Column';
import { ColumnType } from '@core/domain/Model';
import { Expression } from '@core/domain/Model/Column/Expression/Expression';

export class NestedColumn extends Column {
  isNullable: boolean;
  isEncrypted: boolean;
  isPrivate: boolean;
  description?: string;
  className = ColumnType.nested;
  displayName: string;
  name: string;
  nestedColumns: Column[];
  defaultExpression?: Expression;

  constructor(
    displayName: string,
    name: string,
    isNullable: boolean,
    isEncrypted: boolean,
    isPrivate: boolean,
    nestedColumns: Column[],
    description?: string,
    defaultExpression?: Expression
  ) {
    super();
    this.displayName = displayName;
    this.name = name;
    this.nestedColumns = nestedColumns;
    this.description = description;
    this.isNullable = isNullable || false;
    this.isEncrypted = isEncrypted || false;
    this.isPrivate = isPrivate || false;
    this.defaultExpression = defaultExpression;
  }

  static fromObject(obj: NestedColumn): NestedColumn {
    const rawNestedColumns = obj.nestedColumns || [];
    const nestedColumns = rawNestedColumns.map(item => Column.fromObject(item)).filter((item): item is Column => !!item);
    const defaultExpression = obj.defaultExpression ? Expression.fromObject(obj.defaultExpression) : void 0;
    return new NestedColumn(obj.displayName, obj.name, obj.isNullable, obj.isEncrypted, obj.isPrivate, nestedColumns, obj.description, defaultExpression);
  }
}
